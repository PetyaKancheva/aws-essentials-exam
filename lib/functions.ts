import {Duration} from "aws-cdk-lib";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import {Construct} from "constructs";

interface BaseFunctionProps {
    environment: {
        VALID_TOPIC_ARN?: string,
        INVALID_TOPIC_ARN?: string,
        TABLE_NAME?:string,
        DELETION_TOPIC_ARN?:string,
        DELETION_LAMBDA_ARN?:string

    }
}

export class BaseFunction extends NodejsFunction {

    constructor(scope: Construct, id: string, props: BaseFunctionProps) {
        super(scope, id, {
            ...props,
            handler: "handler",
            timeout: Duration.seconds(5),
            entry: `${__dirname}/../src/${id}.ts`,

        })
    }}