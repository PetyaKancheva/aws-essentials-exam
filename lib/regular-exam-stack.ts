import {Stack, StackProps} from 'aws-cdk-lib';
import {LambdaIntegration, RestApi} from 'aws-cdk-lib/aws-apigateway';
import {AttributeType, Table} from 'aws-cdk-lib/aws-dynamodb';
import {Runtime} from 'aws-cdk-lib/aws-lambda';
import {NodejsFunction} from 'aws-cdk-lib/aws-lambda-nodejs';
import {Subscription, SubscriptionProtocol, Topic} from 'aws-cdk-lib/aws-sns';
import {LambdaSubscription} from 'aws-cdk-lib/aws-sns-subscriptions';
import {Endpoint} from 'aws-sdk';
import {Construct} from 'constructs';
import {BaseFunction} from "./functions";
import {AccountPrincipal, Effect, PolicyStatement, PrincipalBase, ServicePrincipal} from 'aws-cdk-lib/aws-iam';

// import * as sqs from 'aws-cdk-lib/aws-sqs';

export class RegularExamStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        // email to be changed by Sisi
        const email = "pe_kan@posteo.com"

        // table
        const table = new Table(this, "SisiTable2", {
            partitionKey: {
                name: "PK",
                type: AttributeType.STRING
            },
            sortKey: {
                name: "SK",
                type: AttributeType.STRING
            }

        });


        // SNS for valid data
        const validDataTopic = new Topic(this, "ValidDataTopic");
        new Subscription(this, "ValidDataSubscription", {
            topic: validDataTopic,
            protocol: SubscriptionProtocol.EMAIL,
            endpoint: email
        });
        // SNS for informing lambda
        const invalidDataTopic = new Topic(this, "InvalidDataTopic");

        // SNS for deletion of item
        const deletionDataTopic = new Topic(this, "DeletionDataTopic");
        new Subscription(this, "DeletionDataSubscription", {
            topic: deletionDataTopic,
            protocol: SubscriptionProtocol.EMAIL,
            endpoint: email
        });
//
        // lambda for validation
        const validationLambda = new BaseFunction(this, "ValidationLambda", {
            environment: {
                VALID_TOPIC_ARN: validDataTopic.topicArn,
                INVALID_TOPIC_ARN: invalidDataTopic.topicArn,
            }
        });
        // grant permittsion to lambda
        validDataTopic.grantPublish(validationLambda);
        invalidDataTopic.grantPublish(validationLambda);

        // labda to delete the item after 24hrs
        const deletionLambda = new BaseFunction(this, "DeletionLambda", {
            environment: {
                TABLE_NAME: table.tableName,
                DELETION_TOPIC_ARN: deletionDataTopic.topicArn,
            }
        });

        // lambda for unvalid JSON -  put in table and crate schedule for deletion

        const invalidDataLambda = new BaseFunction(this, "InvalidDataLambda", {
            environment: {
                TABLE_NAME: table.tableName,
                DELETION_LAMBDA_ARN: deletionLambda.functionArn,
            }
        });
        //
        // // ad policy statement fior klambdy
        // const policyStatement = new PolicyStatement({
        //         effect:  Effect.ALLOW,
        //         resources:["*"],
        //         actions: ['sts:AssumeRole'],
        //         principals: [new ServicePrincipal('scheduler.amazonaws.com')]
        //     });
        //
        //

        invalidDataLambda.addPermission("AllowEventBridgeInvoke",{
            principal: new ServicePrincipal('events.amazonaws.com'),
            action:"lambda:InvokeFunction",
            sourceArn: `arn:aws:events:${this.region}:${this.account}:rule/*`    } );


        invalidDataLambda.addPermission("AllowScheduleInvoke",{
            principal: new ServicePrincipal('scheduler.amazonaws.com'),
            action:"*",
            sourceArn: `arn:aws:scheduler:${this.region}:${this.account}:schedule/default/*`    } );

// get lambda role and add it to environmetn

        const invalidLambdaAssumedRoleRN = invalidDataLambda.role?.roleArn;
        invalidDataLambda.addEnvironment("ROLE_ARN", invalidLambdaAssumedRoleRN!);

        invalidDataTopic.addSubscription(new LambdaSubscription(invalidDataLambda));

        // grant permission to lambda for table
        table.grantFullAccess(invalidDataLambda);


        // grant permission to lambda for table
        table.grantWriteData(deletionLambda);

        // API with resource and  post method
        const sisiApi = new RestApi(this, "SisiApi", {
            restApiName: "SisiApi"
        });
        // resource
        const postResource = sisiApi.root.addResource("order");
        // method post
        postResource.addMethod('POST', new LambdaIntegration(validationLambda,
            {
                proxy: true
            }
        ))

    }
}
