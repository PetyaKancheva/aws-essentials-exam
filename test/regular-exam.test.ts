 

import { App } from "aws-cdk-lib";
import { RegularExamStack } from "../lib/regular-exam-stack";
import { Template } from "aws-cdk-lib/assertions";

// example test. To run these tests, uncomment this file along with the
// example resource in lib/regular-exam-stack.ts
test('SQS Queue Created', () => {
  const app = new App();
    
  const stack = new  RegularExamStack(app, 'MyTestStack');
    
  const template = Template.fromStack(stack);
  
    // 
});
