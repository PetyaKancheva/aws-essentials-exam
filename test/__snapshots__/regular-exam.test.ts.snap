// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Queue Created 1`] = `
{
  "Outputs": {
    "SisiApiEndpointEAD33547": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "SisiApi3319F256",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "SisiApiDeploymentStageprodB39ADFBE",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeletionDataSubscriptionD7222694": {
      "Properties": {
        "Endpoint": "pe_kan@posteo.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "DeletionDataTopic1830E62A",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "DeletionDataTopic1830E62A": {
      "Type": "AWS::SNS::Topic",
    },
    "DeletionLambda1D7334AD": {
      "DependsOn": [
        "DeletionLambdaServiceRoleDefaultPolicyA4294FEA",
        "DeletionLambdaServiceRole1D0EE251",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "62725d9f50e9a1b8d9060caee2eb7d1f6bbb0c47e21f3b4f4cbe997c1b5c71e1.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DELETION_TOPIC_ARN": {
              "Ref": "DeletionDataTopic1830E62A",
            },
            "TABLE_NAME": {
              "Ref": "SisiTable230092D4A",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "DeletionLambdaServiceRole1D0EE251",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DeletionLambdaServiceRole1D0EE251": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DeletionLambdaServiceRoleDefaultPolicyA4294FEA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SisiTable230092D4A",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeletionLambdaServiceRoleDefaultPolicyA4294FEA",
        "Roles": [
          {
            "Ref": "DeletionLambdaServiceRole1D0EE251",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InvalidDataLambdaAllowEventBridgeInvokeDE5798A1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "InvalidDataLambdaC8276619",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:events:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":rule/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "InvalidDataLambdaAllowInvokeMyTestStackInvalidDataTopic5FCFAFB92B8D987E": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "InvalidDataLambdaC8276619",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "InvalidDataTopicF8A94006",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "InvalidDataLambdaAllowScheduleInvoke58678126": {
      "Properties": {
        "Action": "*",
        "FunctionName": {
          "Fn::GetAtt": [
            "InvalidDataLambdaC8276619",
            "Arn",
          ],
        },
        "Principal": "scheduler.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:scheduler:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":schedule/default/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "InvalidDataLambdaC8276619": {
      "DependsOn": [
        "InvalidDataLambdaServiceRoleDefaultPolicy7BF3ED52",
        "InvalidDataLambdaServiceRole1ADB9591",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3c17138fbdf1943c07240fb49e4b70321b46f5d61f4aab62c06c2602862d73dc.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DELETION_LAMBDA_ARN": {
              "Fn::GetAtt": [
                "DeletionLambda1D7334AD",
                "Arn",
              ],
            },
            "ROLE_ARN": {
              "Fn::GetAtt": [
                "InvalidDataLambdaServiceRole1ADB9591",
                "Arn",
              ],
            },
            "TABLE_NAME": {
              "Ref": "SisiTable230092D4A",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "InvalidDataLambdaServiceRole1ADB9591",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "InvalidDataLambdaInvalidDataTopic86540F5D": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "InvalidDataLambdaC8276619",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "InvalidDataTopicF8A94006",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "InvalidDataLambdaServiceRole1ADB9591": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "InvalidDataLambdaServiceRoleDefaultPolicy7BF3ED52": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:*",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SisiTable230092D4A",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "InvalidDataLambdaServiceRoleDefaultPolicy7BF3ED52",
        "Roles": [
          {
            "Ref": "InvalidDataLambdaServiceRole1ADB9591",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InvalidDataTopicF8A94006": {
      "Type": "AWS::SNS::Topic",
    },
    "SisiApi3319F256": {
      "Properties": {
        "Name": "SisiApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "SisiApiAccountDD0D4D12": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "SisiApi3319F256",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "SisiApiCloudWatchRole3B46039D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "SisiApiCloudWatchRole3B46039D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "SisiApiDeployment2550FB0D33f31531dafaf9b74a2ee870054a3c1a": {
      "DependsOn": [
        "SisiApiorderPOST0877EDA4",
        "SisiApiorderDDAA9388",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "SisiApi3319F256",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "SisiApiDeploymentStageprodB39ADFBE": {
      "DependsOn": [
        "SisiApiAccountDD0D4D12",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "SisiApiDeployment2550FB0D33f31531dafaf9b74a2ee870054a3c1a",
        },
        "RestApiId": {
          "Ref": "SisiApi3319F256",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "SisiApiorderDDAA9388": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "SisiApi3319F256",
            "RootResourceId",
          ],
        },
        "PathPart": "order",
        "RestApiId": {
          "Ref": "SisiApi3319F256",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "SisiApiorderPOST0877EDA4": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "ValidationLambda6B331F4E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "SisiApiorderDDAA9388",
        },
        "RestApiId": {
          "Ref": "SisiApi3319F256",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "SisiApiorderPOSTApiPermissionMyTestStackSisiApi9AB9C7A8POSTorder33F6AFFA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ValidationLambda6B331F4E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "SisiApi3319F256",
              },
              "/",
              {
                "Ref": "SisiApiDeploymentStageprodB39ADFBE",
              },
              "/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SisiApiorderPOSTApiPermissionTestMyTestStackSisiApi9AB9C7A8POSTorderAF59CEFD": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ValidationLambda6B331F4E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "SisiApi3319F256",
              },
              "/test-invoke-stage/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SisiTable230092D4A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "ValidDataSubscriptionAA9F5D4D": {
      "Properties": {
        "Endpoint": "pe_kan@posteo.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "ValidDataTopic8A2B0C06",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ValidDataTopic8A2B0C06": {
      "Type": "AWS::SNS::Topic",
    },
    "ValidationLambda6B331F4E": {
      "DependsOn": [
        "ValidationLambdaServiceRoleDefaultPolicyF69C66E3",
        "ValidationLambdaServiceRoleC65F14AB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "8acd953f85ee4a4fddb2c3d3d7f5c249ef685f7e8f9918de1633a8cdd7cf378b.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "INVALID_TOPIC_ARN": {
              "Ref": "InvalidDataTopicF8A94006",
            },
            "VALID_TOPIC_ARN": {
              "Ref": "ValidDataTopic8A2B0C06",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ValidationLambdaServiceRoleC65F14AB",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ValidationLambdaServiceRoleC65F14AB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ValidationLambdaServiceRoleDefaultPolicyF69C66E3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "ValidDataTopic8A2B0C06",
              },
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "InvalidDataTopicF8A94006",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ValidationLambdaServiceRoleDefaultPolicyF69C66E3",
        "Roles": [
          {
            "Ref": "ValidationLambdaServiceRoleC65F14AB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
